internal/component/kafka/consumer.go

```diff
@@ -32,60 +32,63 @@ type consumer struct {
 	b := consumer.k.backOffConfig.NewBackOffWithContext(session.Context())
 	isBulkSubscribe := consumer.k.checkBulkSubscribe(claim.Topic())
-	if isBulkSubscribe {
-		return consumer.processBulkMessages(session, claim, b)
-	}
-	for message := range claim.Messages() {
-		if consumer.k.consumeRetryEnabled {
-			if err := retry.NotifyRecover(func() error {
-				return consumer.doCallback(session, message)
-			}, b, func(err error, d time.Duration) {
-				consumer.k.logger.Warnf("Error processing Kafka message: %s/%d/%d [key=%s]. Error: %v. Retrying...", message.Topic, message.Partition, message.Offset, asBase64String(message.Key), err)
-			}, func() {
-				consumer.k.logger.Infof("Successfully processed Kafka message after it previously failed: %s/%d/%d [key=%s]", message.Topic, message.Partition, message.Offset, asBase64String(message.Key))
-			}); err != nil {
-				consumer.k.logger.Errorf("Too many failed attempts at processing Kafka message: %s/%d/%d [key=%s]. Error: %v.", message.Topic, message.Partition, message.Offset, asBase64String(message.Key), err)
-			}
-		} else {
-			err := consumer.doCallback(session, message)
-			if err != nil {
-				consumer.k.logger.Errorf("Error processing Kafka message: %s/%d/%d [key=%s]. Error: %v.", message.Topic, message.Partition, message.Offset, asBase64String(message.Key), err)
-			}
-		}
-	}
-	return nil
-}
 
-func (consumer *consumer) processBulkMessages(session sarama.ConsumerGroupSession, claim sarama.ConsumerGroupClaim,
-	b backoff.BackOff,
-) error {
-	handlerConfig, err := consumer.k.GetTopicBulkHandlerConfig(claim.Topic())
+	handlerConfig, err := consumer.k.GetTopicHandlerConfig(claim.Topic())
 	if err != nil {
 		return fmt.Errorf("error getting bulk handler config for topic %s: %w", claim.Topic(), err)
 	}
-	ticker := time.NewTicker(time.Duration(handlerConfig.SubscribeConfig.MaxBulkAwaitDurationMilliSeconds) * time.Millisecond)
-	defer ticker.Stop()
-	messages := make([]*sarama.ConsumerMessage, 0, handlerConfig.SubscribeConfig.MaxBulkCount)
-	for {
-		select {
-		case <-session.Context().Done():
-			return consumer.flushBulkMessages(claim, messages, session, handlerConfig.Handler, b)
-		case message := <-(claim.Messages()):
-			if message != nil {
-				messages = append(messages, message)
-				if len(messages) >= handlerConfig.SubscribeConfig.MaxBulkCount {
-					return consumer.flushBulkMessages(claim, messages, session, handlerConfig.Handler, b)
+	if isBulkSubscribe {
+		ticker := time.NewTicker(time.Duration(handlerConfig.SubscribeConfig.MaxBulkAwaitDurationMilliSeconds) * time.Millisecond)
+		defer ticker.Stop()
+		messages := make([]*sarama.ConsumerMessage, 0, handlerConfig.SubscribeConfig.MaxBulkCount)
+		for {
+			select {
+			case <-session.Context().Done():
+				return consumer.flushBulkMessages(claim, messages, session, handlerConfig.BulkHandler, b)
+			case message := <-claim.Messages():
+				if message != nil {
+					consumer.mutex.Lock()
```

Comment:
Shouldn't the lock be placed outside of the `if message != nil` block?

Reply:
Actually, I thought that lock is basically required when accessing `messages` and not `message`. But, it is perfectly fine to be cautious here as its just an if check. Moved it out.